### PR TITLE
Limit variable layer height displaying precision to 3 digits

### PR DESCRIPTION
**Pull request by vovodroid**
Currently in variable layer height slider heights are shown with six digits precision:

image

It has no meaning as Z height is rounded to three digits anyway (and to two digits in slider):

image

This PR rounds variable height calculation and displaying to three digits:

image

### CODE REVIEW

**Comment by igiannakas** [comment]
Personally, as you've already opened the code for this, I would think that rounding the variable layer height to 2 decimals may be preferred as most printers cannot output a 3 digit precision on the z (or any other axis for that matter) in a meaningful way.

For example a Voron 2.4 per full step of the z axis outputs 0.04mm in Z movement without micro stepping, and, assuming a fully precise 16 micro steps, it would output 0.0025mm per micro step. So a 0.001 move is not possible with the standard voron setup for example.

Using 64 micro steps a voron may be able to output a resolution of 0.000625 per micro step so a layer height change of 0.001mm would be 1.6 micro steps to execute which is not really attainable due to micro step drift.

So I'd recommend a 0.01 precision which is achievable with 4 micro steps on the standard config which can be theoretically output with some level of accuracy.

**Comment by vovodroid** [comment]
I use three digits to make it consistent with  Prusa (https://github.com/prusa3d/PrusaSlicer/pull/10298). They took my implementation and limited it to three digits. I have nothing against 0.01.

**Comment by igiannakas** [comment]
Yeap agreed - but I don’t think prusas is right either. The mechanical accuracy of our printers can’t reproduce 3 digits with any meaningful precision. Maybe @SoftFever could chime in here too as I may be wrong :)

**Comment by vovodroid** [comment]
You are right, but for whatever reason g-code resolution is also rounded to  0.001, despite standard XY step is 0.0125

**Comment by igiannakas** [comment]
I thought that XY is quantized by the slicing tolerance value, so if you set 0.0125 there it would create quanta with this irrespective of decimal places 🤔

File: src/libslic3r/Slicing.cpp [hunk 1]
```diff
@@ -314,6 +314,8 @@ std::vector<double> layer_height_profile_adaptive(const SlicingParameters& slici
         else if (layer_height_profile.back() > height && layer_height_profile.back() - height > LAYER_HEIGHT_CHANGE_STEP)
             height = layer_height_profile.back() - LAYER_HEIGHT_CHANGE_STEP;
 
+        height = std::round(height * 1000) / 1000.f;
```
Comment (vovodroid) [inline src/libslic3r/Slicing.cpp:317]: Actually rounding could be skipped.

Comment (igiannakas) [reply src/libslic3r/Slicing.cpp:317]: How so? Dont we still need to round the gcode output, not just the display output?

Comment (vovodroid) [reply src/libslic3r/Slicing.cpp:317]: I don't now. Let's to @SoftFever decide.

**Comment by dewi-ny-je** [comment]
> Personally, as you've already opened the code for this, I would think that rounding the variable layer height to 2 decimals may be preferred as most printers cannot output a 3 digit precision on the z (or any other axis for that matter) in a meaningful way.
> 
> For example a Voron 2.4 per full step of the z axis outputs 0.04mm in Z movement per full step, and, assuming a fully precise 16 micro steps, it would output 0.0025mm per micro step. So a 0.001 move is not possible with the standard voron setup for example.
> 
> Using 64 micro steps a voron may be able to output a resolution of 0.000625 per micro step so a layer height change of 0.001mm would be 1.6 micro steps to execute which is not really attainable due to micro step drift.
> 
> So I'd recommend a 0.01 precision which is achievable with 4 micro steps on the standard config which can be theoretically output with some level of accuracy.

Even better, I would specify the rounding precision as general printer setting: I have a printer which uses T8 rods and 16 microsteps, so 8000/200/16=2.5 microns. sending command with any other quantization will cause a difference between expected and actual layer height, therefore difference in flow rate and Z-related artefacts.

**Comment by vovodroid** [comment]
@dewi-ny-je gcode command rounded to 0.001 anyway.

**Comment by dewi-ny-je** [comment]
> @dewi-ny-je gcode command rounded to 0.001 anyway.

It is also true that with a minimum height of 80 um, a quantization of 2.5 vs commands rounded to 1 results in at most 1.5/80 = ~2% error, hardly noticeable.

File: src/libslic3r/Slicing.cpp (hunk 1 above)
Comment (SoftFever) [reply src/libslic3r/Slicing.cpp:317]: Agreed, rounding here is not necessary.

**Review by SoftFever — APPROVED** [review]
Looks good!
Thank you

**Review by SoftFever — COMMENTED** [review]
Chime in on the discussion.
I vaguely remember SuperSlicer has a setting to define the Z step. I didn't dig into how that parameter is used inside the code.
Adaptive layer height can potentially leverage such a value to limit the step when calculating layer height.
On the other hand, I'm thinking the minimum layer height defined in the extruder setting probably makes more sense as @igiannakas pointed out. One full Z step could be so small that no machine can actually do that consistently.
So, probably the minimum layer height is better.

**Comment by vovodroid** [comment]
According to Chep's test "magic" Z step has no much meaning, unless you use microscope

https://www.youtube.com/watch?v=WIkT8asT90A

**Comment by SoftFever** [comment]
I would be shocked  if it has any impact on any Creality v-wheel printer ;)

**Pull request merged**